### DISCUSSION TITLE
SvelteKit signout error

### DISCUSSION THREAD

**Discussion by LukeKosner** [supabase/supabase#3468] [Questions] [answered]
Hey everyone,

I am building an app utilizing SvelteKit and Supabase. I adapted the [Svelte guide](https://supabase.io/docs/guides/with-svelte) for Svelte Kit. When I run a signOut() function, I get the following error: `Cannot read properties of null (reading 'user')`. I don't know why. Here is the code without the unique app functionality:

signin.svelte:
```
<script>
  import supabase from "$lib/db";
  import { user } from "$lib/ss";
  import { onMount } from "svelte";

  let loading = false;
  let email;

  user.set(supabase.auth.user())
  
  supabase.auth.onAuthStateChange((_, session) => {
    user.set(session.user);
    console.log(supabase.auth.user())
    const id = supabase.auth.user().id
  });

  onMount(async () => {
    const urlSearchParams = new URLSearchParams(
      window.location.hash.substring(1)
    );
    const params = Object.fromEntries(urlSearchParams.entries());
    if (params["error_code"] != null) {
      history.pushState(
        "",
        document.title,
        window.location.pathname + window.location.search
      );
  });

  const handleLogin = async () => {
    try {
      loading = true;
      const { user, session, error } = await supabase.auth.signIn(
        { email },
        { redirectTo: url + "/signin" }
      );
      if (error) throw error;
    } catch (error) {
console.log(error)
    } finally {
      loading = false;
    }
  };

  async function signOut() {
    const user = supabase.auth.user()
    console.log(supabase.auth.user())
   // this works!!
    try {
      loading = true;
      let { error } = await supabase.auth.signOut();
      if (error) throw error;
    } catch (error) {
      console.log(error.message);
    } finally {
      loading = false;
    }
  }
</script>

<svelte:head>
  <title>Sign In | Intersect</title>
</svelte:head>

<div class="container py-4">
  <div class="bg-light p-5 mb-3 rounded-3 rounded-3">

    {#if ...}

      <button on:click|once={signOut} class="btn btn-outline-dark mb-3"
        >Sign Out</button
      >
    {:else}
 
      <form class="row flex flex-center" on:submit|preventDefault={handleLogin}>
        <div class="col-6 form-widget">
          <h2>Sign In with Intersect</h2>
          <p class="description">
            Sign in via magic link with your email below.
          </p>
          <div class="form-floating mb-3">
            <input
              type="email"
              bind:value={email}
              class="form-control"
              id="floatingInput"
              placeholder="name@example.com"
            />
            <label for="floatingInput">Email address</label>
          </div>

          <div>
            <button
              type="submit"
              class="btn btn-outline-dark mb-3"
              disabled={loading}
              >{loading ? "Loading" : "Send magic link"}</button
            >
            <button class="btn mb-3 btn-primary" id="sub" disabled
              >Submit</button
            >
    
          </div>
        </div>
      </form>
    {/if}
  </div>
</div>
```

db.js:

```
import { createClient } from '@supabase/supabase-js'
const supabase = createClient(
   import.meta.env.VITE_SUPABASE_URL,
   import.meta.env.VITE_SUPABASE_ANON_KEY
)
export default supabase
```

ss.js:

```
import { writable } from 'svelte/store';

export const user = writable(false);
```

Hopefully that's enough. I appreciate the help.

**Comment by silentworks** [comment]
Can you state where in the code above the error is coming from, please? It's hard to decipher this as there is plenty of code there, just looking at the code there is more than one reference to user.

**Reply by LukeKosner** [reply]
Hi @silentworks,

The error is coming from a log in an async function. I log user to show that the store is aware of the supabase.auth property.

```
async function signOut() {
    const user = supabase.auth.user()
    console.log(supabase.auth.user())
   // the above works
    try {
      loading = true;
      let { error } = await supabase.auth.signOut();
      if (error) throw error;
    } catch (error) {
      console.log(error.message);
	// the error is coming from this error log
    } finally {
      loading = false;
    }
  }
```

**Reply by silentworks** [reply]
I don't see a log in async function here. I use Svelte kit and you can see how I handle sign out in my project https://github.com/silentworks/lnkd/blob/main/src/lib/common/AppHeader.svelte#L16

**Reply by LukeKosner** [reply]
All functions (or versions of) specified by the Svelte guide by Supabase are present in my original post. Your approach is smart, but very different from the guide. I am trying to see if this is a documentation error or why a simple function would be failing.

**Reply by silentworks** [reply]
You mentioned Svelte kit and not Svelte hence my approach. You can't use the Svelte guide for Svelte kit, they are two different things as it stands.

**Reply by LukeKosner** [reply]
Understood. Is that worth an issue to update the docs with this? Many SvelteKit developers may not know this.

**Comment by evwltrs** [comment]
Did you fix this? Having the same issue with my signout function. I'm using the one provided in the docs. 
This is the func that is throwing the error:
```
async function signOut() {
		try {
			loading = true
			let { error } = await supabase.auth.signOut()
			if (error) throw error
		} catch (error) {
			alert(error.message)
		} finally {
			loading = false
		}
	}
```

Error is `Cannot read properties of null (reading 'user')`

**Reply by LukeKosner** [reply]
Not yet. I have no idea why the error is coming. Do you use SvelteKit or just Svelte? Also, is it worth filing an issue?

**Reply by evwltrs** [reply]
SvelteKit. Some how we need to get it to stop trying to read user before we invoke `auth.signOut()`

**Reply by LukeKosner** [reply]
Agreed. I will tinker to see if I can figure it out.

**Reply by Robert-Nickel** [reply]
I have the same issue now, any recommmendations?

**Reply by malalecherocks** [reply]
you are reading this in your email, and you know the answer, please reply... to help others

**Comment by zakaria-chahboun** [comment]
I fixed it by creating an other route called `logout`,  The route just remove the session from the server side _(destroy the cookie session if you use SSR)_ and also from the client side by `supabase.auth.logout()`.

And finally redirect to the home! If no user/no session exist, Redirect to the home page anyway.

The `logout` route shows only a loading shape in the center of the page! 

You can visit my response in this issue:
https://github.com/supabase/supabase/discussions/2842#discussioncomment-1486453

**Comment by nullart2** [comment, marked as answer]
This error is generated by App.svelte in this part:

```
supabase.auth.onAuthStateChange((_, session) => {
        user.set(session.user)
    })
```

Solution to solve this error:

```
supabase.auth.onAuthStateChange((state, session) => {
    	if (state == 'SIGNED_IN') {
    		user.set(session.user)	
    	} else {
    		user.set(false);
    	}
        
    })
```

**Reply by timsonner** [reply]
This is the fix, nice work.
